fat/fat2/transaction_test.go

```diff
@@ -0,0 +1,317 @@
+		"outputs": [{"type": "pUSD", "amount": 50}]
+	}`,
+}, {
+	Name:  "empty input",
+	Error: "invalid input",
+	TxJSON: `{
+		"input": {},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG", "amount": 50}]
+	}`,
+}, {
+	Name:  "no input address",
+	Error: "invalid input: no address",
+	TxJSON: `{
+		"input": {"type": "PEG", "amount": 50},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG", "amount": 50}]
+	}`,
+}, {
+	Name:  "burn address",
+	Error: "invalid input: FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC",
+	TxJSON: `{
+		"input": {"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG", "amount": 50},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG", "amount": 50}]
+	}`,
+}, {
+	Name:  "no input type",
+	Error: "invalid input: no token type",
+	TxJSON: `{
+		"input": {"address": "FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q", "amount": 50},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG", "amount": 50}]
+	}`,
+}, {
+	Name:  "empty outputs",
+	Error: "at least one output required",
+	TxJSON: `{
+		"input": {"address": "FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q", "type": "PEG", "amount": 50},
+		"outputs": []
+	}`,
+}, {
+	Name:  "conversion to another address",
+	Error: "invalid output at index 0: conversion output address != input address",
+	TxJSON: `{
+		"input": {"address": "FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q", "type": "PEG"},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "pUSD", "amount": 50}]
+	}`,
+}, {
+	Name:  "omitted both input and output amounts",
+	Error: "invalid output at index 0: output amount required if input amount is nil",
+	TxJSON: `{
+		"input": {"address": "FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q", "type": "PEG"},
+		"outputs": [{"address": "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC", "type": "PEG"}]
+	}`,
+}}
+
+// TestTransaction_Valid tests that the Valid function on a Transaction struct is able to catch all formatting errors
+// on a given transaction's input and outputs. This is strictly data-structure validation. Numerical testing
+// (e.g. whether enough inputs are present to cover the outputs) are done in the tests for GetBalanceDeltas
+func TestTransaction_Valid(t *testing.T) {
+	for _, test := range transactionValidationTests {
+		t.Run(test.Name, func(t *testing.T) {
+			assert := assert.New(t)
+			var tx Transaction
+			err := json.Unmarshal([]byte(test.TxJSON), &tx)
+			assert.Nilf(err, "Transaction JSON Unmarshal raised an unexpected error")
+			err = tx.Valid()
+			if len(test.Error) != 0 {
+				if assert.NotNil(err) {
+					assert.Contains(err.Error(), test.Error)
+				}
+				return
+			}
+			require.NoError(t, err)
```

Comment:
useless require. see comments in transactionio_test.go

Reply:
Fixed